gobblin-runtime/src/main/java/gobblin/runtime/app/ServiceBasedAppLauncher.java

```diff
@@ -0,0 +1,242 @@
+  private static final String DEFAULT_APP_SHUTDOWN_TIME_MINUTES = "1";
+
+  /**
+   * A comma separated list of fully qualified classes that implement the {@link Service} interface. These
+   * {@link Service}s will be run in addition to the core services.
+   */
+  public static final String APP_ADDITIONAL_SERVICES = "app.additional.services";
+
+  private static final Logger LOG = LoggerFactory.getLogger(ServiceBasedAppLauncher.class);
+
+  private final int shutdownTime;
+  private final String appId;
+  private final List<Service> services;
+
+  private volatile boolean hasStarted = false;
+  private volatile boolean hasStopped = false;
+
+  private ServiceManager serviceManager;
+
+  public ServiceBasedAppLauncher(Properties properties, String appName) throws Exception {
+    this.shutdownTime =
+        Integer.parseInt(properties.getProperty(APP_SHUTDOWN_TIME_MINUTES, DEFAULT_APP_SHUTDOWN_TIME_MINUTES));
+    this.appId = ApplicationLauncherUtils.newAppId(appName);
+    this.services = new ArrayList<>();
+
+    // Add core Services needed for any application
+    addJobExecutionServerAndAdminUI(properties);
+    addMetricsService(properties);
+    addJMXReportingService();
+
+    // Add any additional Services specified via configuration keys
+    addServicesFromProperties(properties);
+
+    // Add a shutdown hook that interrupts the main thread
+    addInterruptedShutdownHook();
+  }
+
+  /**
+   * Starts the {@link ApplicationLauncher} by starting all associated services. This method also adds a shutdown hook
+   * that invokes {@link #stop()} and the {@link #close()} methods. So {@link #stop()} and {@link #close()} need not be
+   * called explicitly; they can be triggered during the JVM shutdown.
+   */
+  @Override
+  public void start() {
+    if (this.hasStarted) {
+      LOG.warn("ApplicationLauncher has already started");
+      return;
+    } else {
+      this.hasStarted = true;
+    }
+
+    this.serviceManager = new ServiceManager(this.services);
+
+    Runtime.getRuntime().addShutdownHook(new Thread() {
+      public void run() {
+        try {
+          ServiceBasedAppLauncher.this.stop();
+        } catch (ApplicationException e) {
+          LOG.error("Failed to shutdown application", e);
+        } finally {
+          try {
+            ServiceBasedAppLauncher.this.close();
+          } catch (IOException e) {
+            LOG.error("Failed to close application", e);
+          }
+        }
+      }
+    });
+
+    LOG.info("Starting the Gobblin application and all its associated Services");
+
+    // Start the application
+    this.serviceManager.startAsync();
+  }
+
+  /**
+   * Stops the {@link ApplicationLauncher} by stopping all associated services.
+   */
+  @Override
+  public void stop() throws ApplicationException {
+    if (this.hasStopped) {
+      LOG.warn("ApplicationLauncher has already stopped");
+      return;
+    } else {
+      this.hasStopped = true;
```

Comment:
A Guava `Service` cannot be re-started, so the `ServiceBasedAppLauncher` shouldn't be able to restart either, which is why two variables are necessary.